x-pack/legacy/plugins/canvas/public/components/expression_input/expression_input.tsx

```diff
@@ -0,0 +1,297 @@
+          kind: monacoEditor.languages.CompletionItemKind.Field,
+          documentation: { value: getArgReferenceStr(s.argDef), isTrusted: true },
+          insertText: s.text,
+          command: {
+            title: 'Trigger Suggestion Dialog',
+            id: 'editor.action.triggerSuggest',
+          },
+          range: wordRange,
+        };
+      } else if (s.type === 'value') {
+        return {
+          label: s.text,
+          kind: monacoEditor.languages.CompletionItemKind.Value,
+          insertText: s.text,
+          command: {
+            title: 'Trigger Suggestion Dialog',
+            id: 'editor.action.triggerSuggest',
+          },
+          range: wordRange,
+        };
+      } else {
+        return {
+          label: s.fnDef.name,
+          kind: monacoEditor.languages.CompletionItemKind.Function,
+          documentation: {
+            value: getFunctionReferenceStr(s.fnDef),
+            isTrusted: true,
+          },
+          insertText: s.text,
+          command: {
+            title: 'Trigger Suggestion Dialog',
+            id: 'editor.action.triggerSuggest',
+          },
+          range: wordRange,
+        };
+      }
+    });
+
+    return {
+      suggestions,
+    };
+  };
+
+  providerHover = (model: monacoEditor.editor.ITextModel, position: monacoEditor.Position) => {
+    const text = model.getValue();
+    const word = model.getWordAtPosition(position);
+
+    if (!word) {
+      return {
+        contents: [],
+      };
+    }
+
+    const absPosition = model.getValueLengthInRange({
+      startLineNumber: 0,
+      startColumn: 0,
+      endLineNumber: position.lineNumber,
+      endColumn: word.endColumn,
+    });
+
+    const { fnDef, argDef, argStart, argEnd } = getFnArgDefAtPosition(
+      this.props.functionDefinitions,
+      text,
+      absPosition
+    );
+
+    if (argDef && argStart && argEnd) {
+      // Use the start/end position of the arg to generate a complete range to highlight
+      // that includes the arg name and its complete value
+      const startPos = model.getPositionAt(argStart);
+      const endPos = model.getPositionAt(argEnd);
+
+      const argRange = new monacoEditor.Range(
+        startPos.lineNumber,
+        startPos.column,
+        endPos.lineNumber,
+        endPos.column
+      );
+
+      return {
+        contents: [{ value: getArgReferenceStr(argDef), isTrusted: true }],
+        range: argRange,
+      };
+    } else if (fnDef) {
+      return {
+        contents: [
+          {
+            value: getFunctionReferenceStr(fnDef),
+            isTrusted: true,
+          },
+        ],
+      };
+    }
+
+    return {
+      contents: [],
+    };
+  };
+
+  editorWillMount = () => {
+    language.keywords = this.props.functionDefinitions.map(fn => fn.name);
+  };
+
+  render() {
+    const { value, error, fontSize, functionDefinitions, isAutocompleteEnabled } = this.props;
+
+    if (functionDefinitions.length === 0) {
```

Comment:
I need the functionDefinitions so that I can add the function names to the expression language's "keywords" list (that way we get nice syntax highlighting).

Is there a better way to handle this though?

Reply:
Maybe move the language out from the editor, and waaaay high up somewhere (components/app/index?) where we register functions with the interpreter, you grab the functions out and initialize the language.  Would only happen once and then you could forget about passing in all the definitions and instead just pass in the language? Thoughts?

Reply:
I'm gonna give this a shot